Improve implementation

jjerphan opened the pull request:
**Main changes**:
 - Deactivate profiling
 - Use Cython 3 as well as const with fused types memory views
 - Explore compilation flags for vector instructions
 - Modify the conda env specs to build `pdist_agregation` at creation. 
 - Adapt implementation to use a given working memory instead of chunk size 
 - Add GitHub actions

#6 and #8 have been merged in. Comes before #5.

jjerphan committed:
Adapt to use a given working memory

jjerphan committed:
Use const when possible

Cython3 allows using const with fused types memory view.

jjerphan committed:
Revert "New benchmarks: more trials, different chunk sizes"

This reverts commit 8e11101c8f2d83800509be6e7595e6a262f5a709.

jjerphan committed:
Use third optimisation level for compilation

jjerphan committed:
Install Cython 3 via upstream

jjerphan committed:
Revert benchmarks to previous values

jjerphan committed:
Deactivate profiling

jjerphan committed:
Build pdist_agregation when creating the conda env.

jjerphan committed:
Merge branch 'main' into working_memory_adaptation

jjerphan committed:
Add input validation on the new implementation

So that it mimics sklearn's behavior.

ogrisel commented:
Has `profile=True` a significant impact on the performance?

jjerphan commented:
> Has profile=True a significant impact on the performance?

Probably, this is something I am inspecting.

jjerphan committed:
Set the minimum chunk size to 20

jjerphan commented:
Benchmark results as of 7b884f0:

![nearestneighbors@7b884_euclidean distance, dtype=np float64, 10 trials _ n_samples_train=100_n_samples_test=10000_n_features=500_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/117831234-ab034200-b274-11eb-8181-c46c263f780d.png)
![nearestneighbors@7b884_euclidean distance, dtype=np float64, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=50_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/117831240-ab9bd880-b274-11eb-9be0-7af10b36835e.png)
![nearestneighbors@7b884_euclidean distance, dtype=np float64, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=100_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/117831242-ac346f00-b274-11eb-83dd-a88885437a48.png)
![nearestneighbors@7b884_euclidean distance, dtype=np float64, 10 trials _ n_samples_train=10000_n_samples_test=100_n_features=500_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/117831244-ac346f00-b274-11eb-846c-09ab220ad0e2.png)
![nearestneighbors@7b884_euclidean distance, dtype=np float64, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=50_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/117831246-accd0580-b274-11eb-8d97-da54686d315e.png)
![nearestneighbors@7b884_euclidean distance, dtype=np float64, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=100_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/117831248-accd0580-b274-11eb-8f0e-fa709db62641.png)

jjerphan committed:
Fix typo: agregation to aggregation

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Add a minimal README

jjerphan committed:
Merge branch 'cython3' into working_memory_adaptation

jjerphan committed:
Compile for the native architecture and using vectorisation

jjerphan committed:
Add sklearn and daal4py interface in

jjerphan committed:
Update notebook

jjerphan committed:
Assume fewer datastructures in memory

jjerphan committed:
Remove verbose mode from pip installation

jjerphan committed:
Parametrise tests on the ratio of numbers of vectors for both sets

jjerphan committed:
Return the number of samples used per chunk

jjerphan committed:
Update name of CSV

jjerphan committed:
Reduce the number of working memories to test against

jjerphan committed:
fixup! Reduce the number of working memories to test against

jjerphan committed:
fixup! Return the number of samples used per chunk

jjerphan commented:
Benchmarks results as of `ee6d43c`:
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=50_n_neighbors=1 0](https://user-images.githubusercontent.com/13029839/118117680-3ce18b00-b3ec-11eb-983d-532e182dd687.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=50_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/118117682-3e12b800-b3ec-11eb-8497-b2bb43243c16.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=50_n_neighbors=100 0](https://user-images.githubusercontent.com/13029839/118117687-3eab4e80-b3ec-11eb-85c5-a2fcd174383e.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=100_n_neighbors=1 0](https://user-images.githubusercontent.com/13029839/118117691-3fdc7b80-b3ec-11eb-8e2e-39aa7630fc38.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=100_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/118117697-410da880-b3ec-11eb-812f-c03fa3fb1a1c.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=100_n_neighbors=100 0](https://user-images.githubusercontent.com/13029839/118117706-423ed580-b3ec-11eb-8e10-bb8d59e8acec.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=10000_n_samples_test=100_n_features=500_n_neighbors=1 0](https://user-images.githubusercontent.com/13029839/118117711-43700280-b3ec-11eb-9931-6718b4d07665.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=10000_n_samples_test=100_n_features=500_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/118117713-44a12f80-b3ec-11eb-9331-c62b2262b456.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=10000_n_samples_test=100_n_features=500_n_neighbors=100 0](https://user-images.githubusercontent.com/13029839/118117721-4539c600-b3ec-11eb-83d7-4dc1db4a8f37.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=50_n_neighbors=1 0](https://user-images.githubusercontent.com/13029839/118117725-466af300-b3ec-11eb-9692-76490438efcb.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=50_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/118117737-4965e380-b3ec-11eb-8468-80795de0cf48.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=50_n_neighbors=100 0](https://user-images.githubusercontent.com/13029839/118117740-4a971080-b3ec-11eb-9890-9b75d45f1484.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=100_n_neighbors=1 0](https://user-images.githubusercontent.com/13029839/118117746-4bc83d80-b3ec-11eb-8593-bb45348a8fb9.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=100_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/118117750-4cf96a80-b3ec-11eb-9853-02cd5be4f042.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=100_n_neighbors=100 0](https://user-images.githubusercontent.com/13029839/118117756-4e2a9780-b3ec-11eb-90c9-daac75f5daf3.png)

jjerphan commented:
Benchmarks results as of ee6d43c:
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=50_n_neighbors=1 0](https://user-images.githubusercontent.com/13029839/118118319-0e17e480-b3ed-11eb-8020-e1718e40df66.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=50_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/118118322-0eb07b00-b3ed-11eb-810a-20df7c983b0c.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=50_n_neighbors=100 0](https://user-images.githubusercontent.com/13029839/118118327-107a3e80-b3ed-11eb-983a-3a4d63afc00c.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=100_n_neighbors=1 0](https://user-images.githubusercontent.com/13029839/118118333-12dc9880-b3ed-11eb-8e20-692b97bf9996.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=100_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/118118334-13752f00-b3ed-11eb-916f-5bc662e8cc59.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000_n_samples_test=1000000_n_features=100_n_neighbors=100 0](https://user-images.githubusercontent.com/13029839/118118340-14a65c00-b3ed-11eb-9bb1-6274df326c63.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=10000_n_samples_test=100_n_features=500_n_neighbors=1 0](https://user-images.githubusercontent.com/13029839/118118345-153ef280-b3ed-11eb-8e84-2fad89df2e22.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=10000_n_samples_test=100_n_features=500_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/118118348-15d78900-b3ed-11eb-91cd-8bc19ca744c1.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=10000_n_samples_test=100_n_features=500_n_neighbors=100 0](https://user-images.githubusercontent.com/13029839/118118352-1708b600-b3ed-11eb-88f7-2e71760008c3.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=50_n_neighbors=1 0](https://user-images.githubusercontent.com/13029839/118118356-18d27980-b3ed-11eb-96b1-2625f531f216.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=50_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/118118357-196b1000-b3ed-11eb-8bbd-0a023a67cd96.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=50_n_neighbors=100 0](https://user-images.githubusercontent.com/13029839/118118360-1a03a680-b3ed-11eb-945b-3e546e615938.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=100_n_neighbors=1 0](https://user-images.githubusercontent.com/13029839/118118364-1b34d380-b3ed-11eb-8383-f20eb8eaeceb.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=100_n_neighbors=10 0](https://user-images.githubusercontent.com/13029839/118118367-1bcd6a00-b3ed-11eb-9084-a84556fd2e8d.png)
![nearestneighbors@ee6d43cab3_euclidean distance, dtype=np float64, omp_num_threads=4, open_blas_num_threads=1, 10 trials _ n_samples_train=1000000_n_samples_test=1000_n_features=100_n_neighbors=100 0](https://user-images.githubusercontent.com/13029839/118118371-1cfe9700-b3ed-11eb-984c-56ef8fc22311.png)

ogrisel commented on src/setup.py:
```diff
@@ -8,7 +8,14 @@
+            '-fopenmp',
+            '-O3',
+            '-ftree-vectorize',
+            '-march=native',
```
For reference this option does not exist when compiling with clang-11 apparently.

jjerphan replied:
I've managed to compile with clang 12, but it's probably better reverting `march` and `mtune` for portability, yes.

ogrisel commented on src/setup.py:
```diff
@@ -8,7 +8,14 @@
+            '-ftree-vectorize',
+            '-march=native',
+            '-mtune=native',
+            '-mavx2'
```
For reference this will not exist for ARM64 machines.

ogrisel commented on src/setup.py:
```diff
@@ -8,7 +8,14 @@
+        extra_compile_args=[
+            '-fopenmp',
+            '-O3',
+            '-ftree-vectorize',
```
It would be interesting just to try '-03' and '-ftree-vectorize' to see if we get the same performance.

jjerphan replied:
They are being reran with fdc9d31 on a smaller config.

mbatoul committed:
Add workflow to publish benchmark results on gh-pages

mbatoul committed:
Setup pre-commit hook for black formatter

mbatoul committed:
Run `pre-commit run --all-files`

mbatoul committed:
Fix

mbatoul committed:
Add README.md

mbatoul committed:
Resolve conflicts

jjerphan committed:
Tidy-up notebook

Remove previous runs output.
Add threads configuration on figures.
List specifications.

mbatoul committed:
Set MKL_NUM_THREADS=1 and OPENBLAS_NUM_THREADS=1 to avoid oversubscription

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jjerphan committed:
Import missing submodule

jjerphan committed:
Merge pull request #8 from scikit-learn-inria-fondation/setup_action

Setup Github action: Run benchmarks and publish results

jjerphan committed:
Merge branch 'main' into working_memory_adaptation

jjerphan committed:
Remove compilation flags for portability

march and mtune favor the machine architecture
used for the compilation which do not make
the shared object portable.

Moreover, avx2 instructions aren't supported
on all machines.

jjerphan committed:
Benchmark configuration for fdc9d31

jjerphan commented:
Benchmarks results as of fdc9d31:

![index](https://user-images.githubusercontent.com/13029839/118463017-95bc6680-b6ff-11eb-82d2-26ffd46ef52e.png)
![index1](https://user-images.githubusercontent.com/13029839/118463021-9654fd00-b6ff-11eb-9fc3-6897d4b0fc33.png)
![index2](https://user-images.githubusercontent.com/13029839/118463025-9654fd00-b6ff-11eb-90cc-28c6d62b7f31.png)

mbatoul commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,70 @@
+
+on:
+  push:
+    branches: [setup_action]
```
It should be changed to:

```suggestion
    branches: [main]
```

mbatoul commented on benchmarks/benchmark.py:
```diff
@@ -34,28 +34,32 @@
             for k in n_neighbors:
-                chunk_sizes = CHUNK_SIZES if chunk else [0]
-                for chunk_size in chunk_sizes:
+                working_memorys = WORKING_MEMS if chunk else [0]
+                for working_memory in working_memorys:
```
What about this? 😃

```suggestion
                working_memories = WORKING_MEMS if chunk else [0]
                for working_memory in working_memories:
```

jjerphan replied:
Yes definitely, I made a quick automatic replacement there!
Probably `working_memory_range` is more adapted as a wording.

jjerphan replied:
Fixed in f8392dd.

jjerphan committed:
Change on-push  branch to main

Co-authored-by: Mathis Batoul <batoulmathis@gmail.com>

jjerphan committed:
Use proper wording for working memory

Co-authored-by: Mathis Batoul <batoulmathis@gmail.com>

jjerphan commented:
Thanks for your review.
If everything is alright for you @mbatoul, I let you merge.

jjerphan committed:
CI Invalidate the cache on pdist_aggregation changes

`hashFiles('environment.yml') }}` is no longer sufficient: as
`src/_pdist_aggregation.pyx` can change, we also need to invalidate
based on this file as well.
